joomla/joomla-cms
# Routing/Language: Refactoring the language system to simplify the code

The code for the current multi-language system is unnecessarily complex. This PR tries to improve on #5135 and simplify the code further.
### Goal

This PR tries to achieve 4 goals:
1. Remove a few inconsistencies in the code that may have resulted in hard-to-find bugs.
2. Improve performance by moving the language-URL-behavior from several different *HelperRoute classes into the single plugin.
3. Allow the routing code to work with the language data where necessary in order to simplify the code there, too.
4. Allow for correct URLs and moving between languages where this is expected.

Especially the last part should be discussed in this PR. The new code does not force you to always use one specific language, but allows you to switch between them. The behavior should be the following:
- If a user accesses a site for the first time, the plugin looks up if it should simply direct to the default language or try to discover the right language via the browser.
- The language of the page is always determined by the URL.
- On each pageload, a cookie is set with the current language. If a page has no attached language information, the data from the cookie is taken, otherwise the URL overrides that cookie.

This means that a user that surfs a french website and then clicks on a link to a specific english page on the site, the user is not redirected to a french page with the english content inside or an error page or similar.
### How to test

Simply apply the changes included in here and see that the URLs stay the same and the behavior for language handling is the way you expect it.
### Routing changes

The above mentioned routing changes are primarily aimed at easing the code in other parts of Joomla. Currently, a URL is first created via a *HelperRoute class, which replaces the ISO language code from the content item with the corresponding SEF code. The language plugin, which right now is run before every other code in the routing process, takes that SEF code, compares it to the existing SEF codes, looks up the right language for it and then takes the SEF code from that language and adds it as part of the path in the URL. After that, it deletes the language code from the query, so that it is not available to the rest of the routing process anymore. Since the language is an important part of looking up the right Itemid, this means that the lookup of that part has to stay in the *HelperRoute class, instead of moving it to the component router itself.
### Notice

I changed the above description massively on January 3rd, 2015 in order to make it better understandable. Between opening the PR with the first description and this updated description, the code was again extensively changed and thus the comments before https://github.com/joomla/joomla-cms/pull/5140#issuecomment-65280032 are not valid anymore.
